geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/RegionProvider.java

```diff
@@ -172,6 +176,42 @@ public RedisStats getRedisStats() {
     }
   }
 
+  /**
+   * Execute the given Callable in the context of a GemFire transaction. On failure there is no
+   * attempt to retry.
+   */
+  public <T> T executeInTransaction(RedisKey key, Callable<T> callable) {
+    Callable<T> txWrappedCallable = getTxWrappedCallable(callable);
+    return execute(key, txWrappedCallable);
+  }
+
+  /**
+   * Execute the given Callable in the context of a GemFire transaction. On failure there is no
+   * attempt to retry.
+   */
+  public <T> T executeInTransaction(RedisKey key, List<RedisKey> keysToLock, Callable<T> callable) {
+    Callable<T> txWrappedCallable = getTxWrappedCallable(callable);
+    return execute(key, keysToLock, txWrappedCallable);
+  }
+
+  private <T> Callable<T> getTxWrappedCallable(Callable<T> callable) {
+    Callable<T> txWrappedCallable = () -> {
+      T result;
+      try {
+        txManager.begin();
+        result = callable.call();
+        txManager.commit();
+      } finally {
```

Comment:
I think you want this finally block to complete the tx. So move the begin outside the try and then in the finally block you either commit or rollback.
It would look something like this:
boolean success = false;
txManager.begin();
try {
  result = callable.call();
  success = true;
} finally {
  if (success) {
    txManager.commit();
  } else {
    txManager.rollback();
  }
}

Reply:
I used this as a guide for the current code: https://gemfire.docs.pivotal.io/910/geode/developing/transactions/directed_example.html.

Reply:
After our discussion, I've made the changes you suggested.